frame/alliance/src/lib.rs

```diff
@@ -0,0 +1,1130 @@
+	///
+	/// Any founder has a special one-vote veto right to the rule setting.
+	#[pallet::storage]
+	#[pallet::getter(fn rule)]
+	pub type Rule<T: Config<I>, I: 'static = ()> = StorageValue<_, Cid, OptionQuery>;
+
+	/// The current IPFS cids of the announcements.
+	#[pallet::storage]
+	#[pallet::getter(fn announcements)]
+	pub type Announcements<T: Config<I>, I: 'static = ()> =
+		StorageValue<_, BoundedVec<Cid, T::MaxAnnouncementsCount>, ValueQuery>;
+
+	/// Maps member and their candidate deposit.
+	#[pallet::storage]
+	#[pallet::getter(fn deposit_of)]
+	pub type DepositOf<T: Config<I>, I: 'static = ()> =
+		StorageMap<_, Blake2_128Concat, T::AccountId, BalanceOf<T, I>, OptionQuery>;
+
+	/// The current set of candidates.
+	/// If the candidacy is approved by a motion, then it will become an ally member.
+	#[pallet::storage]
+	#[pallet::getter(fn candidates)]
+	pub type Candidates<T: Config<I>, I: 'static = ()> =
+		StorageValue<_, BoundedVec<T::AccountId, T::MaxCandidatesCount>, ValueQuery>;
+
+	/// Maps member type to alliance members, including founder, fellow and ally.
+	/// Founders and fellows can propose and vote on alliance motions,
+	/// and ally can only wait to be elevated to fellow.
+	#[pallet::storage]
+	#[pallet::getter(fn members)]
+	pub type Members<T: Config<I>, I: 'static = ()> = StorageMap<
+		_,
+		Twox64Concat,
+		MemberRole,
+		BoundedVec<T::AccountId, T::MaxMembersCount>,
+		ValueQuery,
+	>;
+
+	/// The members are being kicked out. They can't retire during the motion.
+	#[pallet::storage]
+	#[pallet::getter(fn kicking_member)]
+	pub type KickingMembers<T: Config<I>, I: 'static = ()> =
+		StorageMap<_, Blake2_128Concat, T::AccountId, bool, ValueQuery>;
+
+	/// The current blacklist of accounts. The accounts can't submit candidacy.
+	#[pallet::storage]
+	#[pallet::getter(fn account_blacklist)]
+	pub type AccountBlacklist<T: Config<I>, I: 'static = ()> =
+		StorageValue<_, BoundedVec<T::AccountId, T::MaxBlacklistCount>, ValueQuery>;
+
+	/// The current blacklist of websites.
+	#[pallet::storage]
+	#[pallet::getter(fn website_blacklist)]
+	pub type WebsiteBlacklist<T: Config<I>, I: 'static = ()> =
+		StorageValue<_, BoundedVec<UrlOf<T, I>, T::MaxBlacklistCount>, ValueQuery>;
+
+	#[pallet::call]
+	impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		/// Add a new proposal to be voted on.
+		///
+		/// Requires the sender to be founder or fellow.
+		#[pallet::weight((
+			T::WeightInfo::propose_proposed(
+				*length_bound, // B
+				T::MaxFounders::get(), // X
+				T::MaxFellows::get(), // Y
+				T::MaxProposals::get(), // P2
+			),
+			DispatchClass::Operational
+		))]
+		pub fn propose(
+			origin: OriginFor<T>,
+			#[pallet::compact] threshold: u32,
+			proposal: Box<<T as Config<I>>::Proposal>,
+			#[pallet::compact] length_bound: u32,
+		) -> DispatchResult {
+			let proposor = ensure_signed(origin)?;
+			ensure!(Self::is_votable_member(&proposor), Error::<T, I>::NotVotableMember);
```

Comment:
not quite clear why the condition to propose and vote is the same?

Reply:
why not? similar to council, only council member can propose motion.